### ISSUE TITLE
Multiple ngDroplet instnaces on same page

### ISSUE THREAD

**Issue by zjda** [Wildhoney/ngDroplet#49]
Hi, I wrote a directive with isolated scope to wrap ngDroplet.  When I put 2 instances of my directive in a page, both of them receive '$dropletFileAdded' event if a file is added to one of them.  Based on the 2 parameters passed in the handler (p1 and p2), I cannot figure our how to update my model.  The test files are attached.

The problem can be solved by either including target field information or limiting the event in the $scope.  Any idea will be appreciated.

-ZJ

[test.zip](https://github.com/Wildhoney/ngDroplet/files/141635/test.zip)

**Comment by Smiley01987** [comment]
I have the same problem and i have no idea how to fix this.

**Comment by Wildhoney** [comment]
I've added the ability to define the callbacks via attributes &mdash; which I think will help you folks differentiate between which Droplet directive is making those calls.

```html
<droplet ng-model="interface" on-add="onAdd(file)" />
```

I don't have time to document the changes at the moment, but if you [look at the `scope`](https://github.com/Wildhoney/ngDroplet/blob/master/components/ngDroplet.js#L34) then you'll see there are 4 bindings. Then take a look at the [actual call](https://github.com/Wildhoney/ngDroplet/blob/master/components/ngDroplet.js#L427) to see which arguments are passed :+1: 

I've incremented the patch version, as this shouldn't be a breaking change. Using the attributes on the directive are optional &mdash; and until we test this further &mdash; and add a couple of unit tests &mdash; we'll stick with the `$broadcast` approach.

Could you guys give it a test and see if this fixes what you're trying to achieve, please?

**Comment by Smiley01987** [comment]
@Wildhoney i can't seem to get this to work. Could you provide more information how to implement this please?

**Comment by zjda** [comment]
Thank you so much for your quick update.  After I refreshed my workspace with the your latest code, magic happened:  only target directive received "$dropletFileAdded" event.  In the other word, if I add a file in the first droplet, only the first one receives the event but the second does not.  I thought you did not change the $broadcast, so I have gone back and forth with the old version and the new version of ngDroplet to ensure the $broadcast behavior changed.  The new attribute "onAdd" is also working for me.

**Comment by Smiley01987** [comment]
@zjda care to share how you implemented this please?

**Comment by Wildhoney** [comment]
In my haste this morning, I accidentally broke the backwards compatibility and so have released `v0.5.14` and deprecated `v0.5.13` on npm &mdash; please ignore the `$broadcast` behaviour, as that will function as it always has.

@Smiley01987 try something like, where the function &mdash; `onAdd` &mdash; on your controller/directive could change depending on the Droplet instance you're currently handling.

```html
<droplet ng-model="interface" on-add="onAdd(file)" />
```

Is the attribute approach fit for purpose? Please advise on any other changes you folks would be looking at to satisfy your implementation :+1:

**Comment by Wildhoney** [comment]
We have:

```html
<droplet ng-model="interface" on-add="add(file)" on-delete="remove(file)" on-success="success(response, files)" on-error="error(response)" />
```

**Comment by zjda** [comment]
@Wildhoney: Basically you are going to have attributes mirroring your broadcast events.  If so, you may want to add on-ready to the attribute, which will be used to set allowedExtensions and other settings for each individual droplet in a page.  For example, I may have one droplet for photo and another one for pdf/word resume.

@Smiley01987: as Wildhoney mentioned, you just need to add the attribute and have your hander in your controller.  For example, if you have 2 droplets on you page, you can have 

`<droplet ng-model="interface1" on-add="onAdd1(file)" />
<droplet ng-model="interface2" on-add="onAdd2(file)" />`

In you controller, you can have 2 functions

$scope.onAdd1 = function(file) {...} to do whatever you want when a file is added to the first droplet, and $scope.onAdd2 = function(file) {...} for the second droplet.  Or you can develop your own directive to wrap droplet.

Thanks,
-ZJ

**Comment by Wildhoney** [comment]
@zjda :+1: Added `onLoad`.

I don't intend to keep both approaches, but will see how well the attribute functions are taken and may remove the `$broadcast`s in future.

**Comment by zjda** [comment]
I have tested on-add, on-delete, and on-load:

`<droplet` ng-model="interface" on-add="added(file)" on-delete="removed(file)" on-load="loaded()" class="droplet">...</droplet>`

and they work perfectly for me.  Since I have my own upload method, I am not able to test on-success and on-error.

Again, thank you very much for the update. 
![image](https://cloud.githubusercontent.com/assets/14140244/13308204/35dd903c-db26-11e5-81be-e9da2c337bd4.png)

**Comment by Wildhoney** [comment]
No problem!

Small consideration: do you think we're better off with hyphens in the attribute names (`on-add`) &mdash; which I prefer for readability &mdash; or without &mdash; and thus sticking closer to the HTML convention (`onadd`).

@Smiley01987 Did you have any luck with the attribute approach?

It's also worth pointing out that you can pass through additional arguments to those callbacks &mdash; which will also help if you have a dynamic list of `droplet`s.

```html
<droplet on-add="added(file, 'first')" ng-model="interface" />
<droplet on-add="added(file, 'second')" ng-model="interface" />
```

**Comment by zjda** [comment]
I agree with you.  The following looks better 

`<droplet on-add="added(file, 'first')" ng-model="firstInterface" />`

than

`<droplet onadd="added(file, 'first')" ng-model="firstInterface" />`

**Comment by Smiley01987** [comment]
@Wildhoney sadly it still doesnt work for me. The function doesn't get called when adding a file to the droplet.

Here is my code:

**Directive:**

`.directive('gameImages', function () {
    return {
        restrict: 'E',
        templateUrl: 'partials/directives/game-images.html',
        scope : {
            images: "=",
            type: "=",
            maxImages: "="
        },
        link: function(scope, element, attr){

            scope.$on('$dropletReady', function whenDropletReady() {
                // set the accepted file types
                scope.images.allowedExtensions(['jpg', 'jpeg']);
                scope.images.useArray(true);
            });

            scope.$on('$dropletFileAdded', function whenFileAdded(event, model) {

            });

            scope.added = function(file, type){
                console.log('added' + file + ' ' + type);
            };
        }
    }`

**Template:**

`<div class="well">
    <section class="droplet">
        <section class="toolbar text-center">
            <div class="add-files">
                <span class="button add-files">
                    <span class="glyphicon glyphicon-picture"></span>
                    Drop file to attach, or <span class="browse_file">browse</span>
                </span>
                <droplet-upload-multiple ng-model="images" on-add="added(file, type)" id="type" name="type"></droplet-upload-multiple>
            </div>
        </section>

        <droplet ng-model="images" class="disabled">
            <ul class="files">
                <li ng-repeat="model in images.getFiles(images.FILE_TYPES.VALID)" class="col-md-3 text-center">
                    <droplet-preview ng-model="model" ng-show="model.isImage()"></droplet-preview>

                    <button type="button" ng-click="model.deleteFile()" class="btn btn-xs btn-danger delete">
                        <i class="glyphicon glyphicon-trash"></i>
                        <span>Remove</span>
                    </button>
                </li>
            </ul>
        </droplet>
    </section>
    <small ng-class="{ 'color-red': images.getFiles(images.FILE_TYPES.VALID).length > maxImages }">{{images.getFiles(images.FILE_TYPES.VALID).length}}/{{maxImages}} main prize images</small>
</div>
`

**Comment by Wildhoney** [comment]
@Smiley01987 Where do you add the `<droplet />` node to your directive? Is that in **game-images.html**? If that's the case then in there change it to `<droplet on-add="added(file)" /> ` &mdash; `ngDroplet` will then invoke the `added` function in the given `scope` context.

**Comment by Smiley01987** [comment]
@Wildhoney Yes exactly, in game-image.html. I added my code in my previous post.

**Comment by Wildhoney** [comment]
@Smiley01987 I've [set you up a JSFiddle](https://jsfiddle.net/142voLtk/) which illustrates how to achieve this through the attributes. Please let me know if this doesn't make things clearer.

**Comment by Smiley01987** [comment]
@Wildhoney I tried your code and everything works fine, but as soon as I change the droplet element to droplet-upload-multiple with an ng-model it fails.

**Comment by Smiley01987** [comment]
@Wildhoney I moved everything out of the directive and into plain HTML. I also added the $scope function in my main controller but still nothing.

HTML:
`<droplet-upload-multiple on-loaded="loadedWinnerImages(winnerImages)" on-add="addedWinnerImage(file)" ng-model="winnerImages" id="winnerImages" name="winnerImages"></droplet-upload-multiple>`

Scope functions:

`$scope.addedWinnerImage = function(file){
                console.log('Added: ' + file.file.name);
            };

            $scope.loadedWinnerImages = function() {
                console.log('Loaded!');
            };`

**Comment by zjda** [comment]
@Smiley01987 The attributes should be used for `<droplet>` NOT for `<droplet-upload-multiple>` or `<droplet-upload-single>`:

`<droplet ng-model="interface" on-add="added(file)" on-delete="removed(file)" on-load="loaded()" class="droplet">`

**Comment by Wildhoney** [comment]
@zjda is right. I've [updated the JSFiddle](https://jsfiddle.net/142voLtk/1/) to show you how to tie the two directives together &mdash; `droplet` and `droplet-upload-single`.

**Comment by Smiley01987** [comment]
@Wildhoney & @zjda, thanks a lot guys. Now it works perfectly!
